### DISCUSSION TITLE
Migration to TS

### DISCUSSION THREAD

**Discussion by nicolapalavecino** [panva/paseto#27] [Ideas]
Is there any interest in migrating this library to TS? I know there are existing type definitions in the project and even npm tags this repo as written in TS but as far as I see, all the core logic is written in JS.

I'm personally interested in this project since I like authentication mechanisms a lot and I feel they are very important for secure applications. I'm interested in contributing to this project @panva.

**Comment by panva** [comment]
Migrating to TS for the sake of having TS is in my opinion pointless and a waste of time.

Typescript has been immensely useful in my `jose` project when doing a universal style library, then the build-time type checks are useful to ensure the different runtimes that are being compiled to are the same. I want to do the same for paseto while also bringing in a JS implementation of `v2.public` and `v4.public`. So there's already a universal lib WIP written in TS branch. But it's stuck on implementing `v3.public` because of the lack of support for importing pkcs8 private EC keys without the public key component in webkit and firefox. The implementation is there but it doesn't work in those browsers, and they're important ones so it cannot be left like so. 

For this work to be unblocked what's needed is a pure js implementation of Elliptic curve point multiplication for the P-384 NIST curve. I haven't gotten around to do this without bringing in a ton of dependencies that I just don't want to have a part of the WebAPI build.

**Reply by nicolapalavecino** [reply]
I didn’t listed all the benefits of TS since I supposed it was implicit in the conversation, that’s why I first asked if there was interest in having it written in TS. Didn’t know you had a WIP TS branch, that’s great!

I’ll try to take a look at that algorithm you are mentioning in the branch and see if I can do anything to help out there. Would that be ok?

**Reply by panva** [reply]
Sure. And FWIW i'm aware of the benefits of having a project in TS but I don't see them as that beneficiary neither to myself nor the users if all it's going to do is rewrite the code as-is.
